clarity/src/vm/types/signatures.rs

```diff
@@ -271,1204 +53,180 @@ pub enum FunctionArgSignature {
+            FunctionType::Binary(arg1, arg2, return_type) => {
+                let arg1 = arg1.canonicalize(epoch);
+                let arg2 = arg2.canonicalize(epoch);
+                let return_type = return_type.canonicalize(epoch);
+                FunctionType::Binary(arg1, arg2, return_type)
             }
-        };
-
-        Ok(out)
-    }
-
-    pub fn literal_type_of(x: &Value) -> Result<TypeSignature> {
-        match x {
-            Value::Principal(PrincipalData::Contract(contract_id)) => Ok(CallableType(
-                CallableSubtype::Principal(contract_id.clone()),
-            )),
-            _ => Self::type_of(x),
         }
     }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct FunctionArg {
+    pub signature: TypeSignature,
+    pub name: ClarityName,
+}
 
-    // Checks if resulting type signature is of valid size.
-    pub fn construct_parent_list_type(args: &[Value]) -> Result<ListTypeData> {
-        let children_types: Result<Vec<_>> = args.iter().map(TypeSignature::type_of).collect();
-        TypeSignature::parent_list_type(&children_types?)
+impl From<FixedFunction> for FunctionSignature {
+    fn from(data: FixedFunction) -> FunctionSignature {
+        let FixedFunction { args, returns } = data;
+        let args = args.into_iter().map(|x| x.signature).collect();
+        FunctionSignature { args, returns }
     }
+}
 
-    pub fn parent_list_type(
-        children: &[TypeSignature],
-    ) -> std::result::Result<ListTypeData, CheckErrors> {
-        if let Some((first, rest)) = children.split_first() {
-            let mut current_entry_type = first.clone();
-            for next_entry in rest.iter() {
-                current_entry_type = Self::least_supertype_v2_1(&current_entry_type, next_entry)?;
-            }
-            let len = u32::try_from(children.len()).map_err(|_| CheckErrors::ValueTooLarge)?;
-            ListTypeData::new_list(current_entry_type, len)
+/// Parsing functions.
+pub trait TupleTypeSignatureExt {
+    fn parse_name_type_pair_list<A: CostTracker>(
+        epoch: StacksEpochId,
+        type_def: &SymbolicExpression,
+        accounting: &mut A,
+    ) -> Result<TupleTypeSignature>;
+}
+
+impl TupleTypeSignatureExt for TupleTypeSignature {
+    fn parse_name_type_pair_list<A: CostTracker>(
+        epoch: StacksEpochId,
+        type_def: &SymbolicExpression,
+        accounting: &mut A,
+    ) -> Result<TupleTypeSignature> {
+        if let SymbolicExpressionType::List(ref name_type_pairs) = type_def.expr {
+            let mapped_key_types = parse_name_type_pairs(epoch, name_type_pairs, accounting)?;
+            TupleTypeSignature::try_from(mapped_key_types).map_err(CheckErrors::from)
         } else {
-            Ok(TypeSignature::empty_list())
+            Err(CheckErrors::BadSyntaxExpectedListOfPairs)
         }
     }
 }
 
-/// Parsing functions.
-impl TypeSignature {
+pub trait TypeSignatureExt {
```

Comment:
Please add a docstring explaining this trait.

Reply:
Actually, seeing this massive trait and the traits above it, all of which only seem to exist because `SymbolicExpression` isn't in `clarity-serialization`, I'm inclined to take back what I said earlier about keeping `SymbolicExpression` in `clarity`.  Please instead move `SymbolicExpression` and its ilk to `clarity-serialization` so these traits can be removed.  We'll just port the feature flags from `clarity` to `clarity-serialization`.

Reply:
Agreed with @kantai